client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java

```diff
@@ -309,6 +309,10 @@ public void segmentIsSealed(SegmentIsSealed segmentIsSealed) {
 
         @Override
         public void noSuchSegment(NoSuchSegment noSuchSegment) {
+            if (!state.setupConnection.isReleased()) {
```

Comment:
I think this is problematic because release can become true right after evaluating this predicate, which would invalidate the execution of the block. Is there anything else we can check or perhaps prevent the change of state until we complete the execution of `failConnection`?

Reply:
Yes, `if` the predicate is redundant.
The reason for removing it is as follows.
`NoSuchSegment`  is received either as a response to `SetupAppend` or to a normal `Append`.
- On receiving a `NoSuchSegment` for a `SetupAppend` we can close the connection immediately and invoke the `resendToSuccessorsCallback`.
- On receiving a `NoSuchSegment` for an `Append` we can also close the connection immediately as until the `resendToSuccessorsCallback` is invoked all the writes to this segment will be stored in inflight  and once the callback is invoked we obtain a lock which prevents further writes and all the inflight events are re transmitted.

Reply:
The changes are already in the code. Just wanted to list out the reasons why `if` predicate , as mentioned by you, is incorrect.